Build and test instructions (#469)
stlab/libraries: README.md

## Content

### [Concurrency](https://www.stlab.cc/libraries/concurrency/)
This library provides futures and channels, high level abstractions for implementing algorithms that eases the use of multiple CPU cores while minimizing contention. This library solves several problems of the C++11 and C++17 TS futures.

## Documentation

The complete documentation is available on the [STLab home page](http://stlab.cc).

Release changelogs are listed in [CHANGES.md](CHANGES.md).

## Tested on

- Linux with GCC 11
- Linux with Clang 14
- MacOS with Apple-clang 13.0.1
- Windows with Visual Studio 16

## Requirements

- A standards-compliant C++14, C++17, or C++20 compiler
- **Use with a compiler in C++14-compliant mode** requires Boost.Optional and Boost.Variant >= 1.74.0
- **Building** the library requires CMake 3.23 or later
- **Testing or developing** the library requires Boost.Test >= 1.74.0

## Building

STLab is a standard CMake project. See the [running CMake](https://cmake.org/runningcmake) tutorial
for an introduction to this tool.

### Preparation

1. Create a build directory outside this library's source tree.  In this guide, we'll use a sibling

>    directory called `BUILD`.

Author: I did.  The whole point was to make it stick out like a sore thumb so the user would not miss that they probably want to replace the text with something else.

2. If you are using the library in C++14-compliant mode or need to run the test suite, be sure you
   have the necessary parts of Boost >= 1.74.0 installed.  Linux distributions usually make a
   suitable version available through standard package managers.  On macOS or Linux, you can easilly
   install Boost using [Homebrew](https://brew.sh/).  To install Boost on Windows, you can use
   Boost's [binary installers](https://sourceforge.net/projects/boost/files/boost-binaries/).

3. Install a version of CMake >= 3.23.  If you are on Debian or Ubuntu Linux you may need to use
   `snap` to find one that's new enough.

4. If you are using MSVC, you may need to set environment variables appropriately for your target
   architecture by invoking `VCVARSALL.BAT` with an appropriate option.

### Configure

Run CMake in the root directory of this project, setting `../BUILD` as your build directory.  The
basis of your command will be

```

> cmake -S . -B ../BUILD

Author: I thought about this carefully. It is not so unconventional; I've used CMake for years both ways. It's what our GH actions do.  And as long as this repo is called `libraries`, the user is likely to check it out with a different name, and then these instructions get more complicated or error-prone because they have to refer to a directory whose name we don't know.

> cmake -S . -B ../BUILD

Author: ```suggestion
cmake -S . -B ../BUILD # + other options. See below...
```

> cmake -S . -B ../BUILD

Reviewer: Okay, I don't feel strongly about it.

```

but there are other options you may need to append in order to be successful.  Among them:

> * `-DCMAKE_BUILD_TYPE=`[`Release`|`Debug`] to build the given configuration (required).

Reviewer: ```suggestion
* `-DCMAKE_BUILD_TYPE=`[`Release`|`Debug`] to build the given configuration (required unless you're using visual studio or another multi-config generator).
```

You really want to *not* specify `-DCMAKE_BUILD_TYPE` if you're using a multi-config generator. It doesn't make any sense in that context.

Author: Seriously, it won't be harmless?  Suppose we make Ninja a requirement, just to simplify the overall instructions?

Reviewer: It's probably harmless, but it's misleading. "Hey, I said I want a Release build, why is it building debug binaries?"

Author: I'd rather mislead than have people fail because the instructions were complex. I can't believe CMake isn't designed to let people use it without learning about two kinds of generator. If you use a multi config generator but choose a configuration, it won't just build the configuration you asked for?

Reviewer: > If you use a multi config generator but choose a configuration, it won't just build the configuration you asked for?

No, not in my understanding.

> * `-DSTLAB_USE_BOOST_CPP17_SHIMS=ON` if you are building in C++14 compliant mode.

Reviewer: ```suggestion
* `-DSTLAB_USE_BOOST_CPP17_SHIMS=ON` if you would like to use Boost.Optional and Boost.Variant instead of those in the standard. This is required if your compiler doesn't support the standard versions or if CMAKE_CXX_STANDARD is set to 14.
```

Author: I don't think it's helpful to suggest that people building for C++17 might want to use Boost.Optional and Boost.Variant.  Doesn't that just increase the likelihood of some ODR problem down the road?

> * `-DCMAKE_BUILD_TYPE=`[`Release`|`Debug`] to build the given configuration.

Reviewer: ```suggestion
* `-DCMAKE_BUILD_TYPE=`[`Release`|`Debug`] to build the given configuration. This option is *required* if you're not using a multi-config generator such as Visual Studio or XCode. Otherwise it has no effect.
```

Author: Works for me when I leave it out with `-GNinja`.  Is this extra wrinkle important on some platform?

Reviewer: It is important in the sense as you're getting an unsupported, unspecified configuration. If you do this, you don't get a "Debug" or a "Release" configured build, you get a "" configured build. It may happen to work, but I imagine it is pretty fragile. What the LLVM projected does is [error out](https://reviews.llvm.org/D124153) if you have a "" configured build. We should at the very least not endorse its use and probably do what LLVM does.

Author: Makes total sense; fixing.

* `-DBUILD_TESTING=OFF` if you only intend to build, but not test, this library.

> * `-DSTLAB_NO_STD_COROUTINES=TRUE` if you are targeting an Apple platform.

Reviewer: ```suggestion
* `-DSTLAB_NO_STD_COROUTINES=TRUE` if your compiler doesn't have standard coroutine support. This applies to Apple platforms and when CMAKE_CXX_STANDARD is set to 14 or 17.```

> * `-DBoost_USE_STATIC_LIBS=TRUE` if you will be testing on Windows.

Reviewer: ```suggestion
* `-DBoost_USE_STATIC_LIBS=TRUE` is recommended if you will be testing on Windows.
```

Author: Is it helpful for us to make it optional?  Allowing people to configure in too many different ways is a common pitfall, and since boost libraries will only be linked for testing, I think it's better to keep it simple.

Reviewer: Fine by me.

We also suggest the installation of [Ninja](https://ninja-build.org/) and its use by adding
`-GNinja` to your cmake command line… but ninja is not required.

A typical invocation might look like this:

```

> cmake -S . -B ../BUILD -GNinja -DCMAKE_CXX_STANDARD=17 -DBUILD_TESTING=OFF

Reviewer: This is missing the `-DCMAKE_BUILD_TYPE=Release`.

Author: Good catch.  
```suggestion
cmake -S . -B ../BUILD -GNinja -DCMAKE_CXX_STANDARD=17 -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=OFF
```

```

### Build

If your configuration command was successful, go to your build directory (`cd ../BUILD`) and invoke:

```
cmake --build .
```

## Testing

Running the tests in the `BUILD` directory is as simple as invoking 

```
ctest -C Debug
```
or
```
ctest -C Release
```

> depending on which configuration (`CMAKE_BUILD_TYPE`) you chose to build.

Reviewer: It should be mentioned that the `-C` option only makes sense if someone is not using a multi-config generator.

Author: Do you mean, "*is* using?"  because it seems to work regardless.  I'd rather keep complexity out if possible.

Reviewer: Yes, I mean *is* using.

The documentation for this option doesn't specify what happens when a single-config generator is used with this option.

```
       -C <cfg>, --build-config <cfg>
              Choose configuration to test.

              Some  CMake-generated build trees can have multiple build configurations in the same tree.  This option
              can be used to specify which one should be tested.  Example configurations are Debug and Release.
```

I suppose if it is silently ignored, CMake will never change that unspecified behavior because it would break too many people. We can depend on this through social contract at least :). It makes me nervous that we're depending on unspecified behavior, but, yeah, I think we can, in practice, depend on it. I don't feel strongly about this suggestion.

Author: Counting this as resolved.
